sshd-core/src/main/java/org/apache/sshd/client/auth/pubkey/UserAuthPublicKey.java

```diff
@@ -105,6 +106,19 @@ protected boolean sendAuthDataRequest(ClientSession session, String service) thr
             chosenAlgorithm = null;
         } else if (!currentAlgorithms.isEmpty()) {
             currentAlgorithm = currentAlgorithms.poll();
+            if (chosenAlgorithm != null) {
+                Set<String> knownServerAlgorithms = session.getAttribute(
+                        DefaultClientKexExtensionHandler.SERVER_ALGORITHMS);
+                if (knownServerAlgorithms != null
+                        && knownServerAlgorithms.contains(chosenAlgorithm)) {
+                    // We've tried key 'current' with 'chosenAlgorithm', but it
+                    // failed. However, the server had told us it supported
+                    // 'chosenAlgorithm'. Thus it makes no sense to continue
+                    // with this key and other signature algorithms. Skip to the
+                    // next key, if any.
+                    currentAlgorithm = null;
```

Comment:
Let's add a DEBUG level log message summarizing this behavior so that user can tell when it occurred.

Reply:
Done.